### ISSUE TITLE
error when training using one GPU when multiple GPUs are available

### ISSUE THREAD

**Issue by Liu0329** [roytseng-tw/Detectron.pytorch#54]
I have 1 trivial GPU0 and 4 GPUs (1,2,3,4) in my machine. If I do not specify GPU to use and input:
**python tools/train_net_step.py --dataset coco2017 --cfg configs/e2e_faster_rcnn_R-101-FPN_1x.yaml** 
the error is:
path/miniconda3/lib/python3.6/site-packages/torch/cuda/__init__.py:116: UserWarning:
    Found GPU1 Quadro K600 which is of cuda capability 3.0.
    PyTorch no longer supports this GPU because it is too old.

  warnings.warn(old_gpu_warn % (d, name, major, capability[1]))
INFO train_net_step.py: 361: Training starts !
INFO net.py:  72: Changing learning rate 0.000000 -> 0.006667
Traceback (most recent call last):
  File "tools/train_net_step.py", line 437, in <module>
    main()
  File "tools/train_net_step.py", line 407, in main
    net_outputs = maskRCNN(**input_data)
  File "path/miniconda3/lib/python3.6/site-packages/torch/nn/modules/module.py", line 491, in __call__
    result = self.forward(*input, **kwargs)
  File "path/CODE/Detectron.pytorch/lib/nn/parallel/data_parallel.py", line 82, in forward
    mini_kwargs = dict([(k, v[i]) for k, v in kwargs.items()])
  File "path/CODE/Detectron.pytorch/lib/nn/parallel/data_parallel.py", line 82, in <listcomp>
    mini_kwargs = dict([(k, v[i]) for k, v in kwargs.items()])
IndexError: list index out of range

**Comment by Liu0329** [comment]
I want to specify GPU id when training, the command is
 **CUDA_VISIBLE_DEVICES=2 python tools/train_net_step.py --dataset coco2017 --cfg configs/e2e_faster_rcnn_R-101-FPN_1x.yaml**
In train_net_step.py, I comment the line
`cfg.NUM_GPUS = torch.cuda.device_count()`
to enable the use of "NUM_GPUS: 1" in the yaml file

And I got the following error:
INFO train_net_step.py: 361: Training starts !
INFO net.py:  72: Changing learning rate 0.000000 -> 0.006667
Exception ignored in: <bound method _DataLoaderIter.__del__ of <torch.utils.data.dataloader._DataLoaderIter object at 0x7f229d730ba8>>
Traceback (most recent call last):
  File "path/miniconda3/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 349, in __del__
    self._shutdown_workers()
  File "path/miniconda3/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 328, in _shutdown_workers
    self.worker_result_queue.get()
  File "path/miniconda3/lib/python3.6/multiprocessing/queues.py", line 337, in get
    return _ForkingPickler.loads(res)
  File "path/miniconda3/lib/python3.6/site-packages/torch/multiprocessing/reductions.py", line 70, in rebuild_storage_fd
    fd = df.detach()
  File "path/miniconda3/lib/python3.6/multiprocessing/resource_sharer.py", line 58, in detach
    return reduction.recv_handle(conn)
  File "path/miniconda3/lib/python3.6/multiprocessing/reduction.py", line 182, in recv_handle
    return recvfds(s, 1)[0]
  File "path/miniconda3/lib/python3.6/multiprocessing/reduction.py", line 153, in recvfds
    msg, ancdata, flags, addr = sock.recvmsg(1, socket.CMSG_LEN(bytes_size))
ConnectionResetError: [Errno 104] Connection reset by peer
INFO train_net_step.py: 425: Save ckpt on exception ...
INFO train_net_step.py: 135: save model: Outputs/e2e_faster_rcnn_R-101-FPN_1x/May13-21-57-31_tnt14_step/ckpt/model_step0.pth
INFO train_net_step.py: 427: Save ckpt done.
Traceback (most recent call last):
  File "tools/train_net_step.py", line 407, in main
    net_outputs = maskRCNN(**input_data)
  File "path/miniconda3/lib/python3.6/site-packages/torch/nn/modules/module.py", line 491, in __call__
    result = self.forward(*input, **kwargs)
  File "path/CODE/Detectron.pytorch/lib/nn/parallel/data_parallel.py", line 114, in forward
    return self.gather(outputs, self.output_device)
  File "path/CODE/Detectron.pytorch/lib/nn/parallel/data_parallel.py", line 142, in gather
    return gather(outputs, output_device, dim=self.dim)
  File "path/CODE/Detectron.pytorch/lib/nn/parallel/scatter_gather.py", line 96, in gather
    return gather_map(outputs)
  File "path/CODE/Detectron.pytorch/lib/nn/parallel/scatter_gather.py", line 71, in gather_map
    return {key: gather_map([d[key] for d in outputs]) for key in out}
  File "path/CODE/Detectron.pytorch/lib/nn/parallel/scatter_gather.py", line 71, in <dictcomp>
    return {key: gather_map([d[key] for d in outputs]) for key in out}
  File "path/CODE/Detectron.pytorch/lib/nn/parallel/scatter_gather.py", line 71, in gather_map
    return {key: gather_map([d[key] for d in outputs]) for key in out}
  File "path/CODE/Detectron.pytorch/lib/nn/parallel/scatter_gather.py", line 71, in <dictcomp>
    return {key: gather_map([d[key] for d in outputs]) for key in out}
  File "path/CODE/Detectron.pytorch/lib/nn/parallel/scatter_gather.py", line 65, in gather_map
    return Gather.apply(target_device, dim, *outputs)
  File "path/CODE/Detectron.pytorch/lib/nn/parallel/_functions.py", line 54, in forward
    ctx.input_sizes = tuple(map(lambda i: i.size(ctx.dim), inputs))
  File "path/CODE/Detectron.pytorch/lib/nn/parallel/_functions.py", line 54, in <lambda>
    ctx.input_sizes = tuple(map(lambda i: i.size(ctx.dim), inputs))
RuntimeError: dimension specified as 0 but tensor has no dimensions

**Comment by Liu0329** [comment]
The problem is caused by the new bug in gather(outputs, target_device, dim=0) in nn.parallel.gather in pytorch 0.4.  [https://github.com/pytorch/pytorch/issues/6983]
There are two ways to solve it.
1. Wait for the bug fix in pytorch 0.4
2. unsqueeze the scalar loss to be of 1 dimension, like adding the following in model_builder.py
```
for loss_name, loss in return_dict['losses'].items():
    return_dict['losses'][loss_name] = loss.unsqueeze(0)
return_dict['metrics']['accuracy_cls'] = return_dict['metrics']['accuracy_cls'].unsqueeze(0)
```
